action/rewards/withdraw.go

```diff
@@ -0,0 +1,133 @@
+	}
+	tag2 := kv.Pair{
+		Key:   []byte("tx.Validator"),
+		Value: w.ValidatorAddress.Bytes(),
+	}
+	tags = append(tags, tag, tag2)
+	return tags
+}
+
+func (w Withdraw) Marshal() ([]byte, error) {
+	return json.Marshal(w)
+}
+
+func (w *Withdraw) Unmarshal(bytes []byte) error {
+	return json.Unmarshal(bytes, w)
+}
+
+type withdrawTx struct {
+}
+
+func (withdrawTx) Validate(ctx *action.Context, signedTx action.SignedTx) (bool, error) {
+	withdraw := Withdraw{}
+	err := withdraw.Unmarshal(signedTx.Data)
+	if err != nil {
+		return false, err
+	}
+	err = action.ValidateBasic(signedTx.RawBytes(), withdraw.Signers(), signedTx.Signatures)
+	if err != nil {
+		return false, err
+	}
+
+	err = action.ValidateFee(ctx.FeePool.GetOpt(), signedTx.Fee)
+	if err != nil {
+		return false, err
+	}
+	if !ctx.Validators.IsValidatorAddress(withdraw.ValidatorAddress) {
+		return false, action.ErrInvalidValidatorAddr
+	}
+
+	currency, ok := ctx.Currencies.GetCurrencyByName("OLT")
+	if !ok {
+		panic("no default currency available in the network")
+	}
+	if currency.Name != withdraw.WithdrawAmount.Currency {
+		return false, errors.Wrap(action.ErrInvalidAmount, withdraw.WithdrawAmount.String())
+	}
+	err = withdraw.ValidatorSigningAddress.Err()
+	if err != nil {
+		return false, errors.Wrap(action.ErrInvalidAddress, err.Error())
+	}
+	err = withdraw.ValidatorAddress.Err()
+	if err != nil {
+		return false, errors.Wrap(action.ErrInvalidAddress, err.Error())
+	}
+	return true, nil
+}
+
+func (withdrawTx) ProcessCheck(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	return runWithdraw(ctx, tx)
+}
+
+func (withdrawTx) ProcessDeliver(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	return runWithdraw(ctx, tx)
+}
+
+func (withdrawTx) ProcessFee(ctx *action.Context, signedTx action.SignedTx, start action.Gas, size action.Gas) (bool, action.Response) {
+	return action.BasicFeeHandling(ctx, signedTx, start, size, 1)
+}
+
+var _ action.Msg = &Withdraw{}
+var _ action.Tx = &withdrawTx{}
+
+func runWithdraw(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	withdraw := Withdraw{}
+	err := withdraw.Unmarshal(tx.Data)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, action.ErrUnserializable, withdraw.Tags(), err)
+	}
+
+	//2. Get the difference of amount earned vs amount withdrawn for the validator issuing this transaction
+	//3. Check how much he is eligible to withdraw
+	//4. If the amount withdrawn is less than or equal to amount eligible to be withdrawn, make the transaction success.
+	//5. In case of no failure, add this amount the person withdrew, to total withdrawn amount in cumulative rewards db
+	//maturedBalance, _ := ctx.RewardMasterStore.RewardCm.GetMaturedRewards(withdraw.ValidatorAddress)
+	//fmt.Println("Matured Balance :", maturedBalance, withdraw.ValidatorAddress)
+	err = ctx.RewardMasterStore.RewardCm.WithdrawRewards(withdraw.ValidatorAddress, &withdraw.WithdrawAmount.Value)
```

Comment:
I See that you're only withdrawing from the rewards store. What about the reward Pool?